README
WWW-Scraper-ISBN-Yahoo_Driver
=============================

This module is a Perl extension for ISBN search collection that is provide by
WWW::Scraper::ISBN.

This driver searches for book information from the Yahoo Books online 
catalog.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

This module requires these modules:

  WWW::Mechanize
  WWW::Scraper::ISBN
  WWW::Scraper::ISBN::Record
  WWW::Scraper::ISBN::Driver

For testing the module requires these modules:

  Data::Dumper
  IO::File
  Test::More

For optional testing the module would like these modules:

  Test::CPAN::Meta
  Test::CPAN::Meta::JSON
  Test::Pod
  Test::Pod::Coverage

COPYRIGHT AND LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.


Changes
Revision history for WWW-Scraper-ISBN-Yahoo_Driver
==================================================

0.22    2014-12-14
        - fixed license fields in META.json to be lists.
        - ping fix for Cygwin (thanks to Alexandr Ciornii).
        - revised regexes due to site changes.

0.21    2014-08-13
        - added git repository links to metadata.
        - fixed distribution name in metadata.
        - upgrade to WWW-Scraper-ISBN-1.00.
        - conversion routines now moved to WWW::Scraper::ISBN::Driver.
        - extended test suite.
        - site changed, and even more temperamental than before. Considering
          dumping the driver, as the data returned is minimal, and better
          presented by other sites.
        - reworked test failure checks.

0.20    2013-03-17
        - added isbn-check.pl to examples.
        - change file dates changed to meet W3CDTF standards.
        - META updates.
        - more error reporting in tests.
        - reworked regexs due to site changes.
        - added description field.
        - pages, weight, width and height fields no longer provided.

0.19    2012-05-27
        - spelling fixes. (thanks to Florian Schlichting & Gregor Herrmann).
        - META tests no longer require specific versions.
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.

0.18    2011-03-01
        - upgraded to Test::CPAN::Meta::JSON for tests.

0.17    2010-10-29
        - added further skip test if test books are withdrawn from the site.
        - fixed skip test count.

0.16    2010-10-07
        - amended pingtest() in network tests.

0.15    2010-10-06
        - fixed skip count.
        - provided alternate ping test methods.

0.14    2010-09-15
        - reworked test structure, so easier to add sample sets.

0.13    2010-08-10
        - added better protection from network failures.
        - updated prerequisite requirements.

0.12    2010-08-02
        - added connection protection if client or server offline.
        - documentation updates.

0.11    2010-07-30
        - updates to parser for Yahoo Book site.
        - added fields (if known) for binding, pages, weight, width & height.
        - added ean13, which is the industry name for isbn-13.
        - more tests.
        - removal of Template::Extract in favour of plain regexes.
        - various package file changes (inc adding META.json).

0.10    2009-08-17
        - bugfix to number of tests in skip.

0.09    2009-08-15
        - minor fixes to parsing of Yahoo Books website.
        - added ping tests, to avoid failing when no network connection.
        - real failures produce diagnostics not skips now.
        - added more package tests.

0.08    2008-11-17
        - updated pattern recognition due to page layout changing

0.07    2008-06-27
        - added checks to only run extra tests under AUTOMATED_TESTING
        - switched metatest to use Test-CPAN-Meta
        - upped META.ym spec version and link
        - added isbn10 key into book data
        - fixed missing isbn13 from tests
        - amended tests to search for ISBN10 and ISBN13 values

0.06    2007-07-26
        - Updated README.
        - Corrected Copyright dates in LICENSE file.
        - Updated POD.
        - Added better error reporting.
        - added isbn13 support
        - updated pattern recognition due to page layout changing

0.05    2007-03-01
        - Handwritten META.yml to conform to current specs.
        - Changed all instances of LICENCE to LICENSE.
        - Added META.yml test script.
        - Added LICENSE and Artistic files.
        - Added an examples directory. Nothing in it yet, except a request
          for interesting scripts.
        - POD updates
        - Updated the pattern matching code now that Yahoo have changed their 
          page layouts
        - Devel::Cover = 91.9%

0.04    2001-01-07  
        - handler() moved to WWW::Scraper::ISBN::Driver
        - test suite updated to peform skips if can't access web pages

0.03    2004-08-31
        - Simplified error handling

0.02    2004-04-19
        - Test::More added as a prerequisites for PPMs

0.01    2004-04-10
        - original version


lib/WWW/Scraper/ISBN/Yahoo_Driver.pm
package WWW::Scraper::ISBN::Yahoo_Driver;

use strict;
use warnings;

use vars qw($VERSION @ISA);
$VERSION = '0.22';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::Yahoo_Driver - Search driver for Yahoo Books online catalog.

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from the Yahoo Books online catalog.

=cut

#--------------------------------------------------------------------------

###########################################################################
# Inheritence

use base qw(WWW::Scraper::ISBN::Driver);

###########################################################################
# Modules

use WWW::Scraper::ISBN::Driver;
use WWW::Mechanize;

###########################################################################
# Constants

use constant    YAHOO   => 'http://shopping.yahoo.com';
use constant    SEARCH  => 'https://shopping.yahoo.com/search?fr=yshoppingheader_test2&type=2button&did=0&p=';

#--------------------------------------------------------------------------

###########################################################################
# Public Interface

=head1 METHODS

=over 4

=item C<search()>

Creates a query string, then passes the appropriate form fields to the 
Yahoo Books server.

The returned page should be the correct catalog page for that ISBN. If not the
function returns zero and allows the next driver in the chain to have a go. If
a valid page is returned, the following fields are returned via the book hash:

  isbn          (now returns isbn13)
  isbn10        (no longer provided by Yahoo on page)
  isbn13
  ean13         (industry name)
  title
  author
  pubdate       (no longer provided by Yahoo on page)
  publisher     
  book_link
  image_link
  thumb_link    (same as image_link)
  description
  binding       (if known)
  pages         (no longer provided by Yahoo on page)
  weight        (no longer provided by Yahoo on page)
  width         (no longer provided by Yahoo on page)
  height        (no longer provided by Yahoo on page)

The book_link and image_link refer back to the Yahoo Books website.

=back

=cut

sub search {
    my $self = shift;
    my $isbn = shift;
    $self->found(0);
    $self->book(undef);
    my $data = {};

    # validate and convert into EAN13 format
    my $ean = $self->convert_to_ean13($isbn);
    return $self->handler("Invalid ISBN specified [$isbn]")   
        if(!$ean || (length $isbn == 13 && $isbn ne $ean)
                 || (length $isbn == 10 && $isbn ne $self->convert_to_isbn10($ean)));
    $isbn = $ean;

    my $mech = WWW::Mechanize->new;
    $mech->agent_alias( 'Linux Mozilla' );

    eval { $mech->get( YAHOO . '/books' ) };
    return $self->handler("Yahoo! book website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

	eval { $mech->get( SEARCH . $isbn ) };
    return $self->handler("Yahoo! book search website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());


    # The Results page
    my $content = $mech->content();
    #print STDERR "\n# results=[\n$content\n]\n";

    my ($link,$thumb) = $content =~ m!<div class="img"><a href="([^"]+)"[^>]+><img src="([^"]+)"!is;
    #print STDERR "\n# link=[$link]\n# thumb=[$thumb]\n";

    return $self->handler("Failed to find that book on Yahoo! book website.")
        unless(defined $link);

    $data->{book_link} = $link;

	eval { $mech->get( $data->{book_link} ) };
    return $self->handler("Yahoo! book search website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());


	# The Book page
    my $html = $mech->content();
    #print STDERR "\n# page=[\n$html\n]\n";

	return $self->handler("Could not extract data from Yahoo! result page.")
		unless($html =~ m!</body>\s*</html>!);

#    $data->{thumb_link}     = $thumb;
#    ($data->{image_link})   = $html =~ m!<meta property="og:image" content="([^"]+)"/>!is;
    ($data->{title})        = $html =~ m!<span id="productTitle"[^>]*>([^<]*)</span>!is;
    ($data->{description})  = $html =~ m!<div class="exp-hide"><span itemprop="description">(.*?)</span></div>!is;
    ($data->{publisher})    = $html =~ m!<td class="label"><em>Publisher</em></td><td>([^<]+)</td>!is;
    ($data->{binding})      = $html =~ m!<td class="label"><em>Book Format</em></td><td>([^<]+)</td>!is;
    ($data->{author})       = $html =~ m!<td class="label"><em>Author</em></td><td>([^<]+)</td>!is;
    ($data->{isbn13})       = $ean;
    ($data->{isbn10})       = $self->convert_to_isbn10($ean);

	return $self->handler("Could not extract data from Yahoo! result page.")
		unless(defined $data);

	# trim top and tail
	foreach (keys %$data) { next unless(defined $data->{$_});$data->{$_} =~ s/^\s+//;$data->{$_} =~ s/\s+$//; }

    my $bk = {
        'isbn13'        => $data->{isbn13},
        'isbn10'        => $data->{isbn10},
        'isbn'          => $data->{isbn13},
        'ean13'         => $data->{isbn13},
        'author'        => $data->{author},
        'title'         => $data->{title},
		'book_link'		=> $data->{book_link},
        'image_link'    => $thumb,
        'thumb_link'    => $thumb,
		'description'	=> $data->{description},
        'publisher'     => $data->{publisher},
        'pubdate'       => $data->{pubdate},
		'binding'	    => $data->{binding},
		'pages'		    => $data->{pages},
		'weight'		=> $data->{weight},
		'width'		    => $data->{width},
		'height'		=> $data->{height},
		'depth'		    => $data->{depth},
        'html'          => $html
    };
    $self->book($bk);
    $self->found(1);
    return $self->book;
}

q{currently listening to: 'Drunk In Public' by The Levellers};

__END__

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>,

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-Yahoo_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  Miss Barbell Productions, <http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2004-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut
